--- orig_src/helloworld.java	2009-12-19 01:20:53.000000000 -0800
+++ new_src/helloworld.java	2009-12-19 01:20:53.000000000 -0800
@@ -1,5 +1,16 @@
-class HelloWorld {
-	public static void main(String[] args) {
+/*
+ * The Hello World class.
+ */
+class HelloWorld
+{
+	/*
+	 * The main function in this class.
+	 */
+	public static void main(String[] args)
+	{
+		/*
+		 * Print "Hello world!" to the screen.
+		 */
 		System.out.println("Hello world!");
 	}
 }
