*** orig_src/helloworld.js	2009-12-19 01:20:53.000000000 -0800
--- new_src/helloworld.js	2009-12-19 01:20:53.000000000 -0800
***************
*** 1,13 ****
! function helloWorld() {
      alert("Hello world!");
  }
  
  var data = {
! 	helloWorld2: function() {
  		alert("Hello world!");
  	}
  }
  
! var helloWorld3 = function() {
  	alert("Hello world!");
  }
--- 1,22 ----
! /*
!  * Prints "Hello world!"
!  */
! function helloWorld()
! {
      alert("Hello world!");
  }
  
  var data = {
! 	/*
! 	 * Prints "Hello world!"
! 	 */
! 	helloWorld2: function()
! 	{
  		alert("Hello world!");
  	}
  }
  
! var helloWorld3 = function()
! {
  	alert("Hello world!");
  }
