*** orig_src/helloworld.cs	2009-12-19 01:20:53.000000000 -0800
--- new_src/helloworld.cs	2009-12-19 01:20:53.000000000 -0800
***************
*** 1,5 ****
! public class HelloWorld {
! 	public static void Main() {
  		System.Console.WriteLine("Hello world!");
  	}
  }
--- 1,16 ----
! /*
!  * The Hello World class.
!  */
! public class HelloWorld
! {
! 	/*
! 	 * The main function in this class.
! 	 */
! 	public static void Main()
! 	{
! 		/*
! 		 * Print "Hello world!" to the screen.
! 		 */
  		System.Console.WriteLine("Hello world!");
  	}
  }
